### ISSUE TITLE
(regression) docker_container redeploys at every chef-run if Dockerfiles exposed ports + published ports

### ISSUE THREAD

**Issue by jperville** [someara/chef-docker#514]
A short example:
```
    * docker_container[nginx-sites] action run
      - stopping nginx-sites 
      - deleting nginx-sites
      - update nginx-sites
      -   set exposed_ports to {"80/tcp"=>{}} (was {"443/tcp"=>{}, "80/tcp"=>{}}) # <----- BUG
      - starting nginx-sites
```

The redeploy happens because the cookbook builds the list of exposed ports by only taking into account the `docker_container.port` attribute, while the API build it using multiple sources (including the Dockerfile EXPOSE directives).

Since the cookbook does not know about Dockerfile directives, the exposed_ports is calculated wrong and the container redeploys at each chef-run.

This is a regression of #446 (which I fixed myself in #448 last month).

**Comment by jperville** [comment]
As a workaround, I publish port 443 (to an arbitrary port on the loopback interface) so that all exposed_ports have a port mapping and the container does not redeploy gratuitously anymore.

**Comment by chasebolt** [comment]
these are the lines that are producing this problem. i need to read up on how loading current resources work in 12.5 and make adjustments. i dont think we should be grabbing info from the docker api about a container and setting them as the current resource as the docker api returns merged results (our docker create + Dockerfile). since we are unaware of what is in the Dockerfile our current resource and new resource may not match exactly.

https://github.com/bflad/chef-docker/blob/master/libraries/docker_container.rb#L119-L140

**Comment by jperville** [comment]
In the above example, the load_current_value code would properly read that exposed ports are 80/tcp and 443/tcp (the Dockerfile EXPOSES both ports).

However, the docker_container resource calculates the value of exposed_ports from the value of the `port` attribute (which at best returns a subset of the exposed ports). See https://github.com/bflad/chef-docker/blob/master/libraries/helpers_container.rb#L75-L86

The bug comes because there is no way for the cookbook to know that the port 443 is exposed from the Dockerfile if I don't bind a port to 443 explicitly. There is even a comment on line 76 about "infer `port` from `port_bindings` and `exposed_ports`", which is exactly this issue.

**Comment by chasebolt** [comment]
yea, sadly ive actually been aware of this issue but with the major 2.x change it escaped me (issue closed plus my short term memory == no good) and i havent had time to go back through it. ill cycle back around to it today or tonight and see what can be done.

**Comment by jperville** [comment]
Thank you very much @chasebolt. There is no hurry, because I have a workaround for the moment.

**Comment by someara** [comment]
paging doctor @jkeiser

**Comment by someara** [comment]
Hello! I just attempted to set up conditions to reproduce this here -

https://github.com/bflad/chef-docker/commit/69cc7f8ee6816fa3e358c9ad74a9fbc56374b790

... but it doesn't seem to trigger it.

I converged the resources-182-ubuntu-1504 suite a few times with no luck.
What version of Docker are you on? Are you using the port property or setting exposed_ports directly?

-s

**Comment by someara** [comment]
end of run number two, three, etc

```
Running handlers:
Running handlers complete
Chef Client finished, 0/159 resources updated in 13 seconds
Finished converging <resources-182-ubuntu-1504> (3m9.92s).
```

manual inspection

```
root@resources-182-ubuntu-1504-someara-mwe23spb-laptop:~# docker inspect -f "{{ .Config.ExposedPorts }}" overrides-1
map[4321/tcp:{}]
root@resources-182-ubuntu-1504-someara-mwe23spb-laptop:~# docker inspect -f "{{ .Config.ExposedPorts }}" overrides-2
map[4321/tcp:{} 8877/tcp:{} 9988/tcp:{}]
root@resources-182-ubuntu-1504-someara-mwe23spb-laptop:~#
```

**Comment by jperville** [comment]
Hi @someara, I quite busy this weekend but will reproduce for you in the short term.

Basically I am using a base image which exposes 80 and 443 (eg. phusion/passenger-full), then I make my custom image on top of it, and when deploying I only bind to port 80 (not 443).

**Comment by chasebolt** [comment]
@someara the reason your test case 69cc7f8 isn't working is because the docker_container action is :run_if_missing, adjust this to be :run and your test case will work.

```
         * docker_container[overrides-2] action run
           - deleting overrides-2
           - update overrides-2
           -   set exposed_ports to {"9988/tcp"=>{}, "8877/tcp"=>{}} (was {"4321/tcp"=>{}, "8877/tcp"=>{}, "9988/tcp"=>{}})


         * docker_container[host_override] action create (up to date)
```

**Comment by jperville** [comment]
Thanks @chasebolt for investigating this, I had limited time this weekeend.

**Comment by someara** [comment]
I think I fixed this up in 2.1.3... can you confirm?
Also, what were the other properties that might be in the same situation?
WORKDIR?

We should beef up the tests around these.

-s

**Comment by jperville** [comment]
Confirming that the problem is fixed as of 2.1.3. Thanks @someara.

From my experience, other properties in this situations are VOLUME, and probably WORKDIR too, yes.
